### ISSUE TITLE
OS X slower than Linux, slower than native for short strings

### ISSUE THREAD

**Issue by jakubriedl** [lovell/64#3]
I was benchmarking this library on my laptop and it is more than 2 times slower than native implementation. Is the testing code wrong or the benchmark in readme is wrong.

**HW**
MacBook Pro (15-inch, 2016)
2,7 GHz Intel Core i7
16 GB 2133 MHz LPDDR3

**Code**
```
var now = require('performance-now')
var b64 = require('64')

function run (fn, desc) {
  var start = now()
  for (var i = 0; i < 1000000; i++) {
    fn()
  }
  var end = now()
  console.log(desc, (end - start).toFixed(2), 'ms')
}

console.log('NODE', process.version)
run(() => b64.decode(Buffer.from(b64.encode(Buffer.from('test string')), 'base64')), '64')
run(() => Buffer.from(Buffer.from('test string').toString('base64'), 'base64').toString('ascii'), 'native')
```

**Results**
```
NODE v8.1.4
64     2304.52 ms
native 1058.34 ms
```
```
NODE v7.6.0
64     2644.51 ms
native 1180.36 ms
```
```
NODE v6.10.2
64     2964.18 ms
native 1373.59 ms
```

**Comment by lovell** [comment]
Hello, you may have seen in the README that this module is "Suitable for use with Buffer objects with a length greater than about 5KB."

With very short string lengths, you'll experience the adverse cost of V8's JS to C++ bridge.

What happens if you try your tests with longer input?

**Comment by jakubriedl** [comment]
With 10Kb string it gets better
```
NODE v8.1.4
64.    30883.00 ms
native 35041.08 ms
```

So this is fine. But it should be better tested and described in README. Adding chart would also help

**Comment by lovell** [comment]
I suspect the use of `Buffer.from` in your test cases may be adding a lot of overhead. What happens if you "pre-calculate" the input Buffer's contents?

**Comment by jakubriedl** [comment]
Interesting, after moving the Buffer.from out of test the performance is basically the same
```
NODE v8.1.4
64     13924.76 ms
native 13492.26 ms
```

```
var input = '.....' // 10KB string
var encoded = Buffer.from(input.toString('base64'))
run(() => {
  b64.encode(input)
  b64.decode(encoded)
}, '64')
run(() => {
  input.toString('base64')
  encoded.toString('ascii')
}, 'native')
```

**Comment by jakubriedl** [comment]
I found a bug in last results. Now the results for 20KB are more different, 64 is faster
```
var input = Buffer.from('.....') // 20KB string
var encoded = Buffer.from(input.toString('base64'), 'base64')
run(() => {
  b64.encode(input)
  b64.decode(encoded)
}, '64')
run(() => {
  input.toString('base64')
  encoded.toString('ascii')
}, 'native')
```
```
NODE v8.1.4
64     16098.38 ms
native 23691.19 ms
```

**Comment by lovell** [comment]
The CI builds run [benchmarks](https://github.com/lovell/64/blob/master/test/bench.js) and, thanks to your report, I've noticed there's a ~2x difference between Linux/gcc and OS X/clang on Node 8.

Linux - https://travis-ci.org/lovell/64/jobs/252798828#L1733-L1738
```
Buffer x 67,780 ops/sec ±3.66% (78 runs sampled)
64 x 173,349 ops/sec ±2.97% (79 runs sampled)
Fastest encoder for 10000 bytes is 64

Buffer x 35,903 ops/sec ±5.60% (70 runs sampled)
64 x 154,693 ops/sec ±2.75% (73 runs sampled)
Fastest decoder for 10000 bytes is 64
```
OS X - https://travis-ci.org/lovell/64/jobs/252798830#L1338-L1343
```
Buffer x 62,892 ops/sec ±1.13% (84 runs sampled)
64 x 73,932 ops/sec ±6.82% (62 runs sampled)
Fastest encoder for 10000 bytes is 64

Buffer x 28,104 ops/sec ±7.05% (66 runs sampled)
64 x 84,501 ops/sec ±1.91% (61 runs sampled)
Fastest decoder for 10000 bytes is 64
```
My best guess would be that AVX(2) compilation/detection is not quite working correctly on OS X/clang, or perhaps the OS X build environment is perhaps non-AVX.

Are you able to test `npm install 64 --build-from-source` to compile the bindings locally (rather than rely on prebuilt binaries) to see if this makes things any better for you?

**Comment by jakubriedl** [comment]
with `--build-from-source` the results are basically the same
```
NODE v8.1.4
64    16265.61 ms
native 24128.47 ms
```

**Comment by jakubriedl** [comment]
Maybe it would be great to combine both of these approaches and when buffer is smaller than ~5KB use native implementation and this implementation for bigger buffers. Also to improve developer experience it would be great if library could accept string and create the buffer automatically.

**Comment by lovell** [comment]
Thanks for confirming that `--build-from-source` makes no difference. The OS X discrepancy may be a problem with clang and/or the base64 library code itself.

I quite like the idea of implicitly switching implementations based on input size, plus this is where your suggestion of a graph showing input size against performance would help.

In terms of the API accepting strings, we'd need to ensure that the type-detection overhead doesn't slow things down too much. Happy to accept a PR for this if you're willing and able :)

**Comment by lovell** [comment]
I've updated the upstream base64 library to the latest commit and performance is much improved - see https://github.com/lovell/64#performance

OS X (clang) still doesn't seem to produce quite as efficient code as Linux/Windows but in summary processing Buffers longer than ~300 bytes will benefit from this module.

There is also a threshold for the decision to use this shared library vs Buffer's own implementation - see https://github.com/lovell/64/blob/master/index.js#L7-L8 - thanks for this idea.

This is available via npm as v0.1.2.

Thanks for reporting this and your help debugging.
